exchange-v2/contracts/ExchangeV2Core.sol

```diff
@@ -18,17 +18,87 @@ abstract contract ExchangeV2Core is Initializable, OwnableUpgradeable, AssetMatc
     uint256 private constant UINT256_MAX = 2 ** 256 - 1;
 
     //state of the orders
-    mapping(bytes32 => uint) public fills;
+    mapping(bytes32 => uint) public fills; // take-side fills
+
+    //on-chain orders
+    mapping(bytes32 => OrderAndFee) public onChainOrders;
+
+    //struct to hold on-chain order and its protocol fee, fee is updated if order is updated
+    struct OrderAndFee {
+        LibOrder.Order order;
+        uint fee;
+    }
 
     //events
     event Cancel(bytes32 hash, address maker, LibAsset.AssetType makeAssetType, LibAsset.AssetType takeAssetType);
     event Match(bytes32 leftHash, bytes32 rightHash, address leftMaker, address rightMaker, uint newLeftFill, uint newRightFill, LibAsset.AssetType leftAsset, LibAsset.AssetType rightAsset);
+    event UpsertOrder(LibOrder.Order order);
+    
+    /// @dev Creates new or updates an on-chain order
+    function upsertOrder(LibOrder.Order memory order) external payable {
+        bytes32 orderKeyHash = LibOrder.hashKey(order);
+
+        //checking if order is correct
+        require(_msgSender() == order.maker, "order.maker must be msg.sender");
+        require(order.takeAsset.value > fills[orderKeyHash], "such take value is already filled");
+        
+        uint newTotal = getTotalValue(order, orderKeyHash);
+
+        //value of makeAsset that needs to be transfered with tx 
+        uint sentValue = newTotal;
+
+        if(checkOrderExistance(orderKeyHash)) {
+            uint oldTotal = getTotalValue(onChainOrders[orderKeyHash].order, orderKeyHash);
+
+            sentValue = (newTotal > oldTotal) ? newTotal.sub(oldTotal) : 0;
+
+            //value of makeAsset that needs to be returned to order.maker due to updating of the order
+            uint returnValue = (oldTotal > newTotal) ? oldTotal.sub(newTotal) : 0;
+
+            transferLockedAsset(LibAsset.Asset(order.makeAsset.assetType, returnValue), address(this), order.maker, UNLOCK, TO_MAKER);
```

Comment:
@disenotov : AFAIU, we have locking feature only for ETH_ASSET_CLASS (there is an if-condition below), but here we unlock any asset - is it correct?

Reply:
fixed